Various fixes from LuneOS

Herrie82 opened the pull request:
* Add support for changes introduced in Qt 5.8 and Qt 5.12
* Make systrace optional
* Add rotation support based on sensors
* Fix build with Android 4.4.x

Signed-off-by: Herman van Hazendonk <github.com@herrie.org>

rburchell commented on hwcomposer/hwcomposer.pro:
```diff
@@ -27,7 +27,21 @@ HEADERS += hwcomposer_backend_v10.h
 HEADERS += hwcomposer_backend_v11.h
 
+versionAtLeast(QT_MINOR_VERSION, 8) {
+QT += core-private gui-private egl_support-private waylandcompositor-private dbus sensors fontdatabase_support-private eventdispatcher_support-private theme_support-private
```
please fix the indentation. also, remove 'sensors' here since it's conditionally appended on line 38.

Herrie82 replied:
Adjusted

rburchell commented on hwcomposer/hwcomposer.pro:
```diff
@@ -27,7 +27,21 @@ HEADERS += hwcomposer_backend_v10.h
 
+versionAtLeast(QT_MINOR_VERSION, 8) {
+QT += core-private gui-private egl_support-private waylandcompositor-private dbus sensors fontdatabase_support-private eventdispatcher_support-private theme_support-private
+}
```
} else {

Herrie82 replied:
Adjusted

rburchell commented on hwcomposer/hwcomposer_backend_v11.cpp:
```diff
@@ -72,12 +74,14 @@ struct HwcProcs_v11 : public hwc_procs
 
 static void hwc11_callback_vsync(const struct hwc_procs *procs, int, int64_t)
 {
+#ifdef WITH_SYSTRACE
```
i'd suggest adding static functions wrapping the QSystrace use which can host the ifdef in a single place, rather than littering every function call with it

Tofee replied:
We can also try it this way (pseudo-code):
```c++
#ifndef WITH_SYSTRACE
// Add dummy stub methods for QSysTrace
namespace QSystrace
{
    void begin(const char *module, const char *tracepoint, const char *message, ...) {}
    void end(const char *module, const char *tracepoint, const char *message, ...) {}
    void counter(const char *module, const char *tracepoint, const char *message, ...) {}
};
struct QSystraceEvent {
    QSystraceEvent(const char *module, const char *tracepoint) {}
};
#endif
```

rburchell replied:
Yep, good idea :)

rburchell commented on hwcomposer/qeglfsscreen.cpp:
```diff
@@ -44,11 +44,27 @@
 QEglFSScreen::QEglFSScreen(HwComposerContext *hwc, EGLDisplay dpy)
     : m_hwc(hwc)
     , m_dpy(dpy)
+    #ifdef WITH_SENSORS
```
preprocessor directives should be at column 0

Herrie82 replied:
Adjusted

rburchell commented on hwcomposer/qeglfsscreen.cpp:
```diff
@@ -57,8 +73,22 @@ QEglFSScreen::QEglFSScreen(HwComposerContext *hwc, EGLDisplay dpy)
 
 QEglFSScreen::~QEglFSScreen()
 {
+    #ifdef WITH_SENSORS
```
ditto this block (column 0)

Herrie82 replied:
Adjusted

krnlyng commented on hwcomposer/hwcomposer.pro:
```diff
@@ -29,6 +29,11 @@ HEADERS += hwcomposer_backend_v11.h
 QT += core-private compositor-private gui-private platformsupport-private dbus
 
+enable-sensors {
+    QT += sensors
```
Just out of curiosity, why do we need sensors in the hwcompositor plugin?

Herrie82 replied:
We use it in LuneOS for rotation ;)

rburchell replied:
My guess would be for a different compositor stack, to enable rotation. I'm not sure the QPA plugin would really be the right place to put it architecturally, though.

Herrie82 replied:
We use a different stack in LuneOS v.s. Mer. For our setup this is the most logical place, but we made it optional, so shouldn't impact others that don't enable it :)

Herrie82 commented:
@sledges Any objections to merging this as is? Both LuneOS and AsteroidOS are on newer Qt and could use this in the main branch instead of keeping patches ourselves.

rinigus commented:
Any plans to merge this, would be handy to get higher Qt support and make systrace optional.

I used this PR and made few more changes to bump it to 5.14 in https://github.com/nemomobile-ux/qt5-qpa-hwcomposer-plugin/tree/nemo-dev. Thanks for your work on it!

mlehtima commented on hwcomposer/qeglfsscreen.cpp:
```diff
@@ -57,7 +73,21 @@ QEglFSScreen::QEglFSScreen(HwComposerContext *hwc, EGLDisplay dpy)
 QEglFSScreen::~QEglFSScreen()
 {
+#ifdef WITH_SENSORS
+    if(m_orientationSensor) {
```
A small complaint about formatting, all other files have format "if (something)" so space after if and no spaces around the content inside (). Same switches. A lot of formatting mismatch in the new changes.

mlehtima commented on hwcomposer/hwcomposer_backend_v11.cpp:
```diff
@@ -65,6 +78,7 @@ static qint64 setTime;
 #define QPA_HWC_TIMING_SAMPLE(variable)
 #endif
 
+
```
Random empty line added.

mlehtima replied:
This line is still here, is it intentional?

mlehtima commented on hwcomposer/qeglfsscreen.cpp:
```diff
@@ -44,11 +44,27 @@
+
+#include <QTimer>
+
+#include "moc_qeglfsscreen.cpp"
```
Are moc files added usually like this? I thought those are automatically handled by qmake?

mlehtima commented on hwcomposer/qeglfsscreen.cpp:
```diff
@@ -99,4 +129,40 @@ qreal QEglFSScreen::refreshRate() const
+        m_screenOrientation = screenPrimaryOrientation;
+        break;
+    case QOrientationReading::LeftUp:  /* 90 clockwise */
+        m_screenOrientation = screenPrimaryOrientation == Qt::PortraitOrientation ? Qt::InvertedLandscapeOrientation : Qt::PortraitOrientation;
```
Quite long lines, maybe consider splitting into two lines or so.

mlehtima commented:
Also consider following the commit message formatting used in this repo.

rinigus commented:
Few additional changes that could be useful in this context:

* https://github.com/mer-hybris/qt5-qpa-hwcomposer-plugin/commit/1424e647f7f18ffd0831da9eb51a015dff85abfa
  As hwcomposer/hwcomposer_backend_v20.cpp is also including `<private/qsystrace_p.h>`, I moved your dummy QSystrace   into a separate header and included it in the both backends. In addition, it adds the same platformWindow approach for deliverUpdateRequest.

* https://github.com/mer-hybris/qt5-qpa-hwcomposer-plugin/commit/c96f512fd5409f35bd4525a475028c13f27c76d7 Bumps it to 5.13+ with the replacement for screenAdded/destroyScreen

Herrie82 commented:
@mlehtima I will rework it a bit. Give me a little time for build testing at my end as well.

Herrie82 commented:
@mlehtima @rinigus Updated, squashed some commits, fixed formatting, added the commits mentioned by Ringus as well and updated the commit messages. Builds OK on LuneOS side with Qt 5.15

@mlehtima You might want to have a look at the remaining conflict in the .spec, since this is Mer-build specific.

mlehtima commented:
Would it be possible to rebase the changes in a way to get rid of the merge commit currently in the PR?

mlehtima commented on hwcomposer/qeglfsscreen.cpp:
```diff
@@ -99,4 +128,43 @@ qreal QEglFSScreen::refreshRate() const
+        break;
+    case QOrientationReading::LeftUp:  /* 90 clockwise */
+        m_screenOrientation = screenPrimaryOrientation == Qt::PortraitOrientation ? 
+        Qt::InvertedLandscapeOrientation : Qt::PortraitOrientation;
```
Sorry to nitpick about formatting but maybe add some indentation (add the usual 4 spaces) to the second line in these so it's clear these are part of the previous line.

Herrie82 replied:
Fixed

sledges commented:
> Would it be possible to rebase the changes in a way to get rid of the merge commit currently in the PR?

Could you please also drop the ':' from the commits, `[hwcomposer]: ...` -> `[hwcomposer] ...`

mlehtima commented:
Also something went wrong in rebase, now it shows duplicates of already included changes from master branch.

Herrie82 commented:
> > Would it be possible to rebase the changes in a way to get rid of the merge commit currently in the PR?
> 
> Could you please also drop the ':' from the commits, `[hwcomposer]: ...` -> `[hwcomposer] ...`

Done

Herrie82 commented:
> Also something went wrong in rebase, now it shows duplicates of already included changes from master branch.

@mlehtima Yeah was still working on it, now force pushed a rebased one with just 6 commits. If anything still needs reworking, let me know.
